Deterministic SVG output

ksemb opened the pull request:
SVG output is currently in random order, because some code iterates through unordered dictionaries. The proposed patch iterates through the dictionaries in an order defined by the keys of the dictionary.

tacaswell commented:
Due to #4349 which was causing occasional rendering errors #4388 made it so that svg output will _never_ be fully deterministic.

What is the benefit of doing this?

tacaswell commented:
Can you also rebase your commit on top of current master so that we do not have that merge commit?

ksemb commented:
> Due to #4349 which was causing occasional rendering errors #4388 made it so that svg output will never be fully deterministic.

I also noticed that IDs were random in the current master branch, but I currently work with an older version of matplotlib that has deterministic IDs.

> What is the benefit of doing this?
Deterministic output makes regression testing easier. For instance, if SVG images created with matplotlib are under version control together with the LaTeX sources of a scientific publication and the author makes cosmetic changes to the plotting code, he can use the version control system to check whether or not the SVGs have changed.

> Can you also rebase your commit on top of current master so that we do not have that merge commit?
I rebased the changes on top of the current master branch. The rebased commit has ID
1286886cad0b43e320277fa2135fb1aec3b51810

tacaswell commented on lib/matplotlib/axes/_base.py:
```diff
@@ -37,6 +37,11 @@
 is_string_like = cbook.is_string_like
 is_sequence_of_strings = cbook.is_sequence_of_strings
 
+def _sorted_iteritems(a):
```
I think that the changes in this file are un-related.  This function only get used at call time to generate the arguments passed to the `Line2D` artists from the (kind of terrifying) mess of inputs we support to `plot`.

ksemb replied:
If I remember correctly this file created the black rectangle that surrounds the axes. The rectangle consists of four lines, and without the modifications those four lines were drawn in random order.

tacaswell commented:
Something didn't go quite right with the re-base, all of the merge commits are still there.  After you do the rebase you will need to force-push to what ever branch on gh the PR is tracking (in your case master).  If you are worried about losing work you can always create place-holder branches.

ksemb commented:
I just did the force-push and I think that worked

tacaswell commented on lib/matplotlib/axes/_base.py:
```diff
@@ -176,7 +181,7 @@ def __call__(self, *args, **kwargs):
     def set_lineprops(self, line, **kwargs):
         assert self.command == 'plot', 'set_lineprops only works with "plot"'
-        for key, val in six.iteritems(kwargs):
+        for key, val in _sorted_iteritems(kwargs):
```
This is looping over calling `set_*` methods which only mutates Artists, not adds them.

tacaswell commented on lib/matplotlib/axes/_base.py:
```diff
@@ -185,7 +190,7 @@ def set_lineprops(self, line, **kwargs):
     def set_patchprops(self, fill_poly, **kwargs):
         assert self.command == 'fill', 'set_patchprops only works with "fill"'
-        for key, val in six.iteritems(kwargs):
+        for key, val in _sorted_iteritems(kwargs):
```
Ditto here, this change is just setting properties on existing patch objects.

tacaswell commented on lib/matplotlib/axes/_base.py:
```diff
@@ -2013,7 +2018,7 @@ def draw(self, renderer=None, inframe=False):
         # objects so that the spine has the opportunity to update them.
         if not (self.axison and self._frameon):
-            for spine in six.itervalues(self.spines):
+            for spine in _sorted_itervalues(self.spines):
```
This is looping to over items to remove them from the artist list, this is not needed

tacaswell commented on lib/matplotlib/axes/_base.py:
```diff
@@ -3195,7 +3200,7 @@ def get_children(self):
         children.extend(self.texts)
         children.extend(self.artists)
-        children.extend(six.itervalues(self.spines))
+        children.extend(_sorted_itervalues(self.spines))
```
I suspect this one is needed.

ksemb replied:
Now that the test case is implemented it looks like this is the only change that is required in this file.

tacaswell commented on lib/matplotlib/backends/backend_svg.py:
```diff
@@ -68,6 +68,12 @@
 # OF THIS SOFTWARE.
 # --------------------------------------------------------------------
 
+def _sorted_iteritems(a):
```
Can you move these into `matplotlib.cbook` so we don't have multiple definitions of them floating around?

tacaswell replied:
This also needs a docstring.

tacaswell commented on lib/matplotlib/backends/backend_svg.py:
```diff
@@ -238,7 +244,7 @@ def generate_transform(transform_list=[]):
     if attrib:
         output = io.StringIO()
-        attrib = list(six.iteritems(attrib))
+        attrib = list(_sorted_iteritems(attrib))
```
This already gets sorted in the next line.

tacaswell commented on lib/matplotlib/backends/backend_svg.py:
```diff
@@ -923,7 +929,7 @@ def _draw_text_as_path(self, gc, x, y, s, prop, angle, ismath, mtext=None):
             if glyph_map_new:
                 writer.start('defs')
-                for char_id, glyph_path in six.iteritems(glyph_map_new):
+                for char_id, glyph_path in _sorted_iteritems(glyph_map_new):
```
Where does `char_id` come from?  Are these deterministic/have a meaningful order?

tacaswell commented on lib/matplotlib/backends/backend_svg.py:
```diff
@@ -68,6 +68,12 @@
 # --------------------------------------------------------------------
 
+def _sorted_iteritems(a):
+    return ((k,v) for k,v in sorted(six.iteritems(a)))
```
This is sorting on the `(k, v)` tuple, not just the key.

It might be better to do `return ((k, a[k]) for k in sorted(a))`

ksemb replied:
The lines `attrib.sort()` in the same file also sort the tuples...

tacaswell commented:
First, I am coming across in this PR as way more hostile than I am/intend to be, sorry about that.  Thank you for your work and tolerance of terse code review :sheep: .

I am convinced this is a good idea, we just have to be careful about how we document/advertise it (we don't want to pin our selves to an _exact_ vector output as it would make re-organizing the file in a way that preserves what is rendered (ie this PR!) a breaking change) and to get the change right so we don't impact more code paths than needed.

Do you have any ideas on how to test this in such a way that it does not get inadvertently broken, but does not fall in to the trap described above?  Anything I can think of is very non-trivial.

I am also pretty sure that this is only an issue on versions of python where the salt in the dict hash is enabled (you can manually turn it on back to 2.6, but is on by default in > 3.4).  A detail of the cpython implementation is that the keys are yielded in the order of the hashes of their keys which _is_ deterministic, just not the same as their natural sorting.  This leads to a DOS attack vector as if you know how to generate hash collisions you can construct keys which will screw with a python-based web-app (for example).  Hence, the salt was added added (and turned on by default in 3.4) and hence the flurry of bug reports we have had related to non-deterministic dictionary iteration order.

ksemb commented:
I think it would be nice to have the following test case to check the intended behaviour:
1. Save the same figure to 100 SVGs
2. Check if the contents of the first SVG are the same as the contents of the other 99 SVGs

With a test case one could then remove all unnecessary sorting calls.

However, as you already noted, this will not work until now because
> Due to #4349 which was causing occasional rendering errors #4388 made it so that svg output will never be fully deterministic.

To get fully deterministic SVG output the salt generation has to be modified first. What about allowing the user to explicitly specify the salt using a configuration option? Then, when the IDs are deterministic, a test case can be implemented for the current PR.

tacaswell commented:
That will not work, the sorting is always fixed within a given python
process. To test if the order is stable you will need to spawn a new
process (I got bit by this trying to debug another bug).

The salt in the hashes is solving a major issue with browsers rendering so
that needs to stay (at least by default). The tests should be clever enough
to ignore the ids (probably just hard code to ignore the clip details?).

On Mon, May 18, 2015, 05:15 ksemb <notifications@github.com> wrote:

> I think it would be nice to have the following test case to check the
> intended behaviour:
> 1. Save the same figure to 100 SVGs
> 2. Check if the contents of the first SVG are the same as the contents of
> the other 99 SVGs
>
> With a test case one could then remove all unnecessary sorting calls.
>
> However, as you already noted, this will not work until now because
>
> Due to #4349 <https://github.com/matplotlib/matplotlib/issues/4349> which
> was causing occasional rendering errors #4388
> <https://github.com/matplotlib/matplotlib/pull/4388> made it so that svg
> output will never be fully deterministic.
>
> To get fully deterministic SVG output the salt generation has to be
> modified first. What about allowing the user to explicitly specify the salt
> using a configuration option? Then, when the IDs are deterministic, a test
> case can be implemented for the current PR.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/4434#issuecomment-102986487>
> .
>

ksemb commented:
> To test if the order is stable you will need to spawn a new process (I got bit by this trying to debug another bug).

That is good to know. It should however not be too hard to create a separate process for each draw command using the multiprocessing module.
> The salt in the hashes is solving a major issue with browsers rendering so that needs to stay (at least by default).

If it is enabled by default but can be deactivated by the user then that is OK for my use case
> The tests should be clever enough to ignore the ids

For me the motivation for deterministic output is SVG images that are under version control. The version control system will also not be clever enough to ignore the IDs.

ksemb commented:
I have now added two more commits.

The commit "optional user defined svg salt" adds an option "svg.salt" which, when set to a value that is not the empty string, replaces uuid4() as the salt.

The commit "test svg determinism" adds a test case that compares the SVG output for different python hash seeds. This is done using the multiprocessing module in combination with the environment variable "PYTHONHASHSEED".

tacaswell commented on lib/matplotlib/axes/_base.py:
```diff
@@ -834,7 +839,7 @@ def cla(self):
         self.xaxis.cla()
         self.yaxis.cla()
-        for name, spine in six.iteritems(self.spines):
+        for name, spine in _sorted_iteritems(self.spines):
```
This is not needed as it is looping over the spins to clear them.

tacaswell commented on lib/matplotlib/backends/backend_svg.py:
```diff
@@ -136,7 +142,7 @@ def start(self, tag, attrib={}, **extra):
             attrib = attrib.copy()
             attrib.update(extra)
-            attrib = list(six.iteritems(attrib))
+            attrib = list(_sorted_iteritems(attrib))
```
This is already sorted in the next line.

tacaswell commented:
Something is not right with the tests.

Can you please remove the un-needed sorts and add an entry to https://github.com/matplotlib/matplotlib/blob/master/doc/users/whats_new/rcparams.rst for this rcparam.

This also needs some prose documentation someplace.  The level of deterministic output we should guarantee is that with the same mpl micro version the same code will get an identical svg, but no guarantee between versions.

ksemb commented:
> Something is not right with the tests.

The new test requires `multiprocessing.get_context`, which is only available since Python 3.4. The test is now disabled for older Python versions.

> Can you please remove the un-needed sorts and add an entry to https://github.com/matplotlib/matplotlib/blob/master/doc/users/whats_new/rcparams.rst for this rcparam.

I have removed the un-needed sorts and added an entry to `rcparams.rst`.

> This also needs some prose documentation someplace. The level of deterministic output we should guarantee is that with the same mpl micro version the same code will get an identical svg, but no guarantee between versions. 

Where is that place and who should write the documentation?

tacaswell commented on lib/matplotlib/backends/backend_svg.py:
```diff
@@ -16,7 +16,7 @@
      FigureManagerBase, FigureCanvasBase
 from matplotlib.backends.backend_mixed import MixedModeRenderer
-from matplotlib.cbook import is_string_like, is_writable_file_like, maxdict
+from matplotlib.cbook import is_string_like, is_writable_file_like, maxdict, sorted_iteritems, sorted_itervalues
```
This file is not pep8 checked apparently.

Can you write this is

```python
from matplotlib.cbook import (is_string_like, is_writable_file_like, maxdict, 
                              sorted_iteritems, sorted_itervalues)
```

tacaswell commented on lib/matplotlib/rcsetup.py:
```diff
@@ -820,6 +820,7 @@ def __call__(self, s):
     # True to save all characters as paths in the SVG
     'svg.embed_char_paths': [True, deprecate_svg_embed_char_paths],
     'svg.fonttype':         ['path', validate_svg_fonttype],
+    'svg.hashsalt':         [None, validate_any],
```
To match the documentation this should validate as string or None.

tacaswell replied:
Can you also add the rcparam to the template file in the top-level directory?

tacaswell commented on lib/matplotlib/cbook.py:
```diff
@@ -2422,3 +2422,44 @@ def get_instancemethod(self):
+    q.close()
+    return res
+
+def map_random_hash_seeds(f, iterable):
```
Can this be moved to the testing file?  It seem very specific to this test

tacaswell commented:
This is converging nicely.  Left a few minor comments all style/doc relate

jenshnielsen commented on lib/matplotlib/tests/test_backend_svg.py:
```diff
@@ -107,6 +108,43 @@ def test_bold_font_output():
+    # If the Python version is below 3.4, do not run the test.
+    import sys
+    if sys.version_info[:2] < (3, 4):
+        return
```
You should probably Raise a skip test here rather than just returning true

See https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/tests/test_rcparams.py#L111

ksemb commented:
I have now included the suggestions by @tacaswell and @jenshnielsen

tacaswell commented:
Can you clone of the backend _*.rst files in https://github.com/matplotlib/matplotlib/tree/master/doc/api to add and svg one (you will also need to add a line in https://github.com/matplotlib/matplotlib/blob/master/doc/api/index_backend_api.rst).

Along with the auto-generated API docs that is where the description of the deterministic vs not deterministic output should go.  They should at mention the two use cases (which are rather at odds with each other).

ksemb commented:
> Can you clone of the backend _*.rst files in https://github.com/matplotlib/matplotlib/tree/master/doc/api to add and svg one (you will also need to add a line in https://github.com/matplotlib/matplotlib/blob/master/doc/api/index_backend_api.rst).

Yes, I have done that now.
> Along with the auto-generated API docs that is where the description of the deterministic vs not deterministic output should go. They should at mention the two use cases (which are rather at odds with each other).

Apart from the auto-generated docs I could not find any documentation in the existing `rst` files. Can you point me to an exemplary documentation file?

tacaswell commented:
@ksemb This needs a rebase.

WeatherGod commented:
For some reason, the Travis test suite keeps timing out for py34.

tacaswell commented:
If it was the added `sort`s I would expect all of them to be timing out instead of just one.

ksemb commented:
I have now found the reason for the timeout, but have not yet managed to fix it. The timeout happened because a worker processes crashed without informing the master thread that it had crashed. The master process therefore waited forever for the results of the worker to arrive.

The updated code now uses `multiprocess.map` instead of the previous custom implementation for parallel processing. `multiprocess.map` correctly handles exceptions thrown by worker processes. In the new implementation it is therefore possible to see what error actually happened in the worker process. The error raised is`ImportError: cannot import name 'ft2font'`

According to http://sourceforge.net/p/matplotlib/mailman/message/5557051/ this happens if matplotlib is started from the matplotlib root directory. What I do not understand is why the error shows up in the worker threads, but not while launching the master thread.

tacaswell commented:
@ksemb Am I understanding correctly that on all of the other versions this is silently failing and that the multi-processing code here is the code in nose, not in mpl?

tacaswell commented:
Given tn he discussion about dropping 2.6 in the very near future, we may want to hold off on this until then so that we can use ordereddicts instead of sorting in access.

ksemb commented:
>Am I understanding correctly that on all of the other versions this is silently failing and that the multi-processing code here is the code in nose, not in mpl?

The code that is failing is line with context.Pool(processes=1) as pool:

tacaswell commented:
The problem may be that nose is already running in mulit-process mode.

ksemb commented:
I replaced `multiprocessing.map` with `subprocess.check_call([sys.executable, ....])` and the test for determinism passes now.

tacaswell commented:
I would still like to redo this using ordereddicts instead of sorting the
keys, but we need to drop 2.6 first

On Mon, Sep 28, 2015, 05:59 ksemb <notifications@github.com> wrote:

> I replaced multiprocessing.map with subprocess.check_call([sys.executable,
> ....]) and the test for determinism passes now.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/4434#issuecomment-143698013>
> .
>

QuLogic commented:
Master has now dropped 2.6; @ksemb can you rebase and try out `OrderedDict`s instead?

ksemb commented:
> Master has now dropped 2.6; @ksemb can you rebase and try out OrderedDicts instead?

The commits are now rebased and the sorting is replaced by `OrderedDict`s

mdboom commented:
I wonder if this will fix the SVG failures I'm seeing in #5307 (one can hope).  I probably won't have time to dig into that for the next few days, but I have no problem with this being merged whether it helps with that or not...

I restarted the failing test -- it seems like a random unrelated failure.

mdboom commented on lib/matplotlib/axes/_base.py:
```diff
@@ -17,6 +17,7 @@
 from matplotlib import cbook
 from matplotlib.cbook import (_check_1d, _string_to_bool, iterable,
                               index_of, get_label)
+from collections import OrderedDict
```
I know we don't follow this much yet, but can you move stdlib imports above the local (matplotlib) imports?

QuLogic replied:
To be clear, it should be somewhere in the block at lines 7-10 (for standard library imports).

mdboom commented on lib/matplotlib/backends/backend_svg.py:
```diff
@@ -17,6 +17,7 @@
      FigureManagerBase, FigureCanvasBase
 from matplotlib.backends.backend_mixed import MixedModeRenderer
 from matplotlib.cbook import is_string_like, is_writable_file_like, maxdict
+from collections import OrderedDict
```
Ditto here about ordering imports.

QuLogic replied:
And this one on line 8 (which should be broken up, I'd guess, eventually).

tacaswell commented:
:+1: 

@ksemb Thanks for bearing with us through an excessively long review process.

mdboom commented:
This did turn out to be a key piece of solving #5307, so it is incorporated there (I cherry-picked the commit from here, so when #5307 is merged, this should automatically get merged as well).

mdboom commented:
This was included as part of #5307 and merged there.

ksemb commented:
> This did turn out to be a key piece of solving #5307, so it is incorporated there (I cherry-picked the commit from here, so when #5307 is merged, this should automatically get merged as well).

Will the test case and the deterministic hash seed also be merged?

QuLogic commented:
Your other work was applied in #5671.
